worlds/lingo/static_logic.py

```diff
@@ -0,0 +1,372 @@
+import os
+import yaml
+
+from typing import Dict, NamedTuple, Optional, List
+
+
+class RoomAndDoor(NamedTuple):
+    room: Optional[str]
+    door: str
+
+
+class RoomAndPanel(NamedTuple):
+    room: Optional[str]
+    panel: str
+
+
+class RoomEntrance(NamedTuple):
+    room: str  # source room
+    door: Optional[RoomAndDoor]
+    painting: bool
+
+
+class Room(NamedTuple):
+    name: str
+    entrances: List[RoomEntrance]
+
+
+class Door(NamedTuple):
+    name: str
+    item_name: str
+    location_name: Optional[str]
+    panels: Optional[List[RoomAndPanel]]
+    skip_location: bool
+    skip_item: bool
+    door_ids: List[str]
+    painting_ids: List[str]
+    event: bool
+    group: Optional[str]
+    include_reduce: bool
+    junk_item: bool
+
+
+class Panel(NamedTuple):
+    required_rooms: List[str]
+    required_doors: List[RoomAndDoor]
+    colors: List[str]
+    check: bool
+    event: bool
+    internal_ids: List[str]
+    exclude_reduce: bool
+    achievement: bool
+
+
+class Painting(NamedTuple):
+    id: str
+    room: str
+    enter_only: bool
+    exit_only: bool
+    orientation: str
+    required: bool
+    required_when_no_doors: bool
+    required_door: Optional[RoomAndDoor]
+    disable: bool
+    move: bool
+
+
+class Progression(NamedTuple):
+    item_name: str
+    index: int
+
+
+class StaticLingoLogic:
+    ROOMS: Dict[str, Room] = {}
+    PANELS: Dict[str, Panel] = {}
+    DOORS: Dict[str, Door] = {}
+    PAINTINGS: Dict[str, Painting] = {}
+
+    ALL_ROOMS: List[Room] = []
+    DOORS_BY_ROOM: Dict[str, Dict[str, Door]] = {}
+    PANELS_BY_ROOM: Dict[str, Dict[str, Panel]] = {}
+    PAINTINGS_BY_ROOM: Dict[str, List[Painting]] = {}
+
+    PROGRESSIVE_ITEMS: List[str] = []
+    PROGRESSION_BY_ROOM: Dict[str, Dict[str, Progression]] = {}
+
+    PAINTING_ENTRANCES: int = 0
+    PAINTING_EXITS: int = 0
+    REQUIRED_PAINTING_ROOMS: List[str] = []
+    REQUIRED_PAINTING_WHEN_NO_DOORS_ROOMS: List[str] = []
+
+    def __init__(self):
+        try:
+            from importlib.resources import files
+        except ImportError:
+            from importlib_resources import files
+
+        with files("worlds.lingo").joinpath("LL1.yaml").open() as file:
```

Comment:
If it works when I test from the source tree (not as an apworld), shouldn't it work in a release as well? What would I have to do to test that it would work in a non-dev environment?

This line is the only bit of code I've found so far that works both in the source tree and as an apworld.

Reply:
> Shouldn't it work in a release as well?
Yes, if it's on the "apworld exclude list" in Autoworld.py, meaning it will get compiled into a folder in lib/worlds with .pyc files in it. 

But if your question is "Does it mean it will automatically work as an .apworld", the answer to that is no.

To test it, you have to run setup.py --build-exe (I think), with Lingo not on the apworld exclude list, make sure that lib/worlds contains lingo.apworld instead of a lingo folder, and then test a yaml the way you usually would.

The problem arises because .apworld is essentially a .zip, and open() doesn't work for files inside a .zip.

Reply:
Here is the apworld specification for some additional information:
https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/apworld%20specification.md

Note that .apworld compatibility is not currently required but heavily encouraged. And, like I said, you would have to put your game on this list if it doesn't work as an .apworld: 
https://github.com/ArchipelagoMW/Archipelago/blob/main/setup.py#L63-L91